apm-collector/apm-collector-agentstream/src/main/java/org/skywalking/apm/collector/agentstream/worker/segment/buffer/OffsetManager.java

```diff
@@ -0,0 +1,152 @@
+import org.skywalking.apm.collector.core.util.Const;
+import org.skywalking.apm.collector.core.util.TimeBucketUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author pengys5
+ */
+public enum OffsetManager {
+    INSTANCE;
+
+    private final Logger logger = LoggerFactory.getLogger(OffsetManager.class);
+
+    private static final String OFFSET_FILE_PREFIX = "offset";
+    private File offsetFile;
+    private Offset offset;
+    private boolean initialized = false;
+    private RandomAccessFile randomAccessFile = null;
+    private String lastOffsetRecord = Const.EMPTY_STRING;
+
+    public synchronized void initialize() throws IOException {
+        if (!initialized) {
+            this.offset = new Offset();
+            File dataPath = new File(SegmentBufferConfig.BUFFER_PATH);
+            if (dataPath.mkdirs()) {
+                createOffsetFile();
+            } else {
+                File[] offsetFiles = dataPath.listFiles(new PrefixFileNameFilter());
+                if (CollectionUtils.isNotEmpty(offsetFiles) && offsetFiles.length > 0) {
+                    for (int i = 0; i < offsetFiles.length; i++) {
+                        if (i != offsetFiles.length - 1) {
+                            offsetFiles[i].delete();
+                        } else {
+                            offsetFile = offsetFiles[i];
+                        }
+                    }
+                } else {
+                    createOffsetFile();
+                }
+            }
+            String offsetRecord = FileUtils.INSTANCE.readLastLine(offsetFile);
+            offset.deserialize(offsetRecord);
+            initialized = true;
+
+            Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(() -> flush(), 10, 3, TimeUnit.SECONDS);
+        }
+    }
+
+    private void createOffsetFile() throws IOException {
+        String timeBucket = String.valueOf(TimeBucketUtils.INSTANCE.getSecondTimeBucket(System.currentTimeMillis()));
+        String offsetFileName = OFFSET_FILE_PREFIX + "_" + timeBucket + "." + Const.FILE_SUFFIX;
+        offsetFile = new File(SegmentBufferConfig.BUFFER_PATH + offsetFileName);
+        this.offset.getWriteOffset().setWriteFileName(Const.EMPTY_STRING);
+        this.offset.getWriteOffset().setWriteFileOffset(0);
+        this.offset.getReadOffset().setReadFileName(Const.EMPTY_STRING);
+        this.offset.getReadOffset().setReadFileOffset(0);
+        this.flush();
+    }
+
+    public void flush() {
+        String offsetRecord = offset.serialize();
+        if (!lastOffsetRecord.equals(offsetRecord)) {
+            if (offsetFile.length() >= BufferFileConfig.BUFFER_OFFSET_MAX_FILE_SIZE) {
+                exchangeFile();
+            }
+            FileUtils.INSTANCE.writeAppendToLast(offsetFile, randomAccessFile, offsetRecord);
+            lastOffsetRecord = offsetRecord;
+        }
+    }
+
+    private void exchangeFile() {
```

Comment:
I think your name should be `nextFile()`, right? Exchange means using one file to get other file.

Reply:
For data file, your name is `newDataFile`.